### PR TITLE
LibGfx/JPEGXL: Add a JPEG-XL decoder :^) 

### PR DESCRIPTION
**Pull request by LucasChollet**
JPEG-XL is a new image format standardized by the same committee as the
original JPEG image format. It has all the nice feature of recent
formats, and great compression ratios. For more details, look at:
https://jpegxl.info/

This decoder is far from being feature-complete, as it features a grand
total of 60 FIXMEs and TODOs, but anyway, it's still a good start.

I developed this decoder in the Serenity way, I just try to decode a
specific image while staying as close as possible to the specification.

Considering that the format supports a lot of options, and that we
basically support only one possibility for each of them, I'm pretty sure
that we can only decode the image I've developed this decoder for.

Which is:
0aff 3ffa 9101 0688 0001 004c 384b bc41
5ced 86e5 2a19 0696 03e5 4920 8038 000b

The image in a spoofed `.txt` to get Github to accept it (31 B)
[art.jxl.txt](https://github.com/SerenityOS/serenity/files/11853316/art.jxl.txt)

The equivalent PNG (6.4 MB)
art_png

https://twitter.com/jonsneyers/status/1649687661999685632?s=20

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,2006 @@
...
+ */
+
+#include <AK/BitStream.h>
+#include <AK/Endian.h>
+#include <AK/FixedArray.h>
+#include <AK/String.h>
+#include <LibGfx/ImageFormats/JPEGXLLoader.h>
+
+namespace Gfx {
+
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp:14]: having a link to a spec doc up here would be nice

**Review by nico — COMMENTED** [review]
That's a lot of code :)

Looks vaguely right-shaped; I wonder if we have to reimplement all the entropy decoding (see below). Other than that, seems fine.

File: Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,2006 @@
...
+{
+    m_context = make<JPEGXLLoadingContext>(move(stream));
+}
+
+JPEGXLImageDecoderPlugin::~JPEGXLImageDecoderPlugin() = default;
+
+IntSize JPEGXLImageDecoderPlugin::size()
+{
+    if (m_context->state() < JPEGXLLoadingContext::State::HeaderDecoded)
+        m_context->decode_image_header().release_value_but_fixme_should_propagate_errors();
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp:1932]: `ImageDecoderPlugin::size()` should have the right signature in the first place!
This issue exists in every decoder, and yes we should do something about it, but that seems unrelated to this PR.

Probably something I can fix along #19815.

File: Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp [hunk 3]
```diff
@@ -0,0 +1,2006 @@
...
+        configs[0] = TreeConfig { .bits = 0, .from = 0, .to = 1, .offset = 0 };
+
+        Vector<u16> symbols;
+        TRY(symbols.try_resize(1));
+        symbols[0] = symbol;
+
+        return Histogram { configs, symbols };
+    };
+
+    static ErrorOr<Histogram> from_code_lengths(Vector<u8, 18> const& code_lengths)
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp]: I totally forgot about that :sweat_smile:, it is exactly Brotli (apart a unique special case)

When I first faced it in the JPEG-XL spec, I gave a look to `BrotliDecompressionStream` but figured out that it does way more stuff that I needed and that I can't use the public interface. So I decided to implement it and get back to it if the decoder reached a "working" state. But I forgot the last part.

Thank you for your comment!

File: Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp (hunk 1 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/JPEGXLLoader.cpp:15]: As the spec is closed, no public link is accessible. It's not solved, but there is nothing we can't do about it.

**Review by ADKaster — APPROVED** [review]
Definitely agree with nico on the "vaguely right shaped" part :^). Looks neato. Once there's less TODOs, we should add a fuzzer.

**Comment by LucasChollet** [comment]
A fuzzer and a bazillion of tests :yakstack:!
